OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/base/BaseNfcActivity.java

```diff
@@ -105,6 +107,83 @@ public void handleNfcError(IOException e) {
 
         Log.e(Constants.TAG, "nfc error", e);
         Notify.create(this, getString(R.string.error_nfc, e.getMessage()), Style.WARN).show();
+    }
+
+    public void handleNfcError(CardException e) {
+        Log.e(Constants.TAG, "card error", e);
+
+        short status = e.getResponseCode();
+        // When entering a PIN, a status of 63CX indicates X attempts remaining.
+        if ((status & (short)0xFFF0) == 0x63C0) {
+            Notify.create(this, getString(R.string.error_pin, status & 0x000F), Style.WARN).show();
+            return;
+        }
+
+        // Otherwise, all status codes are fixed values.
+        switch (status) {
+            // These errors should not occur in everyday use; if they are returned, it means we
+            // made a mistake sending data to the card.
+            case 0x6A80:
+                throw new AssertionError("Card returned 'Wrong Data' status; this is a programming error!");
+            case 0x6883:
+                throw new AssertionError("Card expected last command in a chain; this is a programming error!");
+            case 0x6B00:
+                throw new AssertionError("Card reported invalid P1/P2 parameter; this is a programming error!");
+            case 0x6D00:
+                throw new AssertionError("Instruction (INS) not supported by smart card; this is a programming error!");
+            case 0x6E00:
+                throw new AssertionError("Class (CLA) not supported by smart card; this is a programming error!");
```

Comment:
This is a good point; I made this assumption from reading through the ykneo-openpgp source code, where they shouldn't happen unless we messed up; but a misbehaving card could force a crash by returning one of these statuses. Plus, someone developing a new NFC-compatible thing could mistakenly crash the app when they're testing it, and an alert would be more useful in that case. I'll turn them into snackbar notifications.